lib/backend/k8s/resources/ipam_block.go

```diff
@@ -0,0 +1,246 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resources
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
+	"github.com/projectcalico/libcalico-go/lib/net"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	IPAMBlockResourceName = "IPAMBlocks"
+	IPAMBlockCRDName      = "ipamblocks.crd.projectcalico.org"
+)
+
+func NewIPAMBlockClient(c *kubernetes.Clientset, r *rest.RESTClient) K8sResourceClient {
+	// Create a resource client which manages k8s CRDs.
+	rc := customK8sResourceClient{
+		clientSet:       c,
+		restClient:      r,
+		name:            IPAMBlockCRDName,
+		resource:        IPAMBlockResourceName,
+		description:     "Calico IPAM blocks",
+		k8sResourceType: reflect.TypeOf(apiv3.IPAMBlock{}),
+		k8sResourceTypeMeta: metav1.TypeMeta{
+			Kind:       apiv3.KindIPAMBlock,
+			APIVersion: apiv3.GroupVersionCurrent,
+		},
+		k8sListType:  reflect.TypeOf(apiv3.IPAMBlockList{}),
+		resourceKind: apiv3.KindIPAMBlock,
+	}
+
+	return &ipamBlockClient{rc: rc}
+}
+
+// Implements the api.Client interface for IPAMBlocks.
+type ipamBlockClient struct {
+	rc customK8sResourceClient
+}
+
+func (c ipamBlockClient) toV1(kvpv3 *model.KVPair) *model.KVPair {
+	cidrStr := kvpv3.Value.(*apiv3.IPAMBlock).Annotations["projectcalico.org/cidr"]
+	_, cidr, err := net.ParseCIDR(cidrStr)
+	if err != nil {
+		panic(err)
+	}
+
+	ab := kvpv3.Value.(*apiv3.IPAMBlock)
+
+	// The CRD needs ints, not *int, so convert here.
+	allocations := []*int{}
+	for _, a := range ab.Spec.Allocations {
+		if a == -1 {
+			allocations = append(allocations, nil)
+		} else {
+			a2 := a
+			allocations = append(allocations, &a2)
+		}
+	}
+
+	// Convert attributes.
+	attrs := []model.AllocationAttribute{}
+	for _, a := range ab.Spec.Attributes {
+		attrs = append(attrs, model.AllocationAttribute{
+			AttrPrimary:   a.AttrPrimary,
+			AttrSecondary: a.AttrSecondary,
+		})
+	}
+
+	return &model.KVPair{
+		Key: model.BlockKey{
+			CIDR: *cidr,
+			UID:  &ab.UID,
+		},
+		Value: &model.AllocationBlock{
+			CIDR:           *cidr,
+			Affinity:       ab.Spec.Affinity,
+			StrictAffinity: ab.Spec.StrictAffinity,
+			Allocations:    allocations,
+			Unallocated:    ab.Spec.Unallocated,
+			Attributes:     attrs,
+			Deleting:       ab.Spec.Deleting,
+		},
+		Revision: kvpv3.Revision,
+	}
+}
+
+func (c ipamBlockClient) v3Fields(k model.Key) (name, cidr string) {
+	cidr = fmt.Sprintf("%s", k.(model.BlockKey).CIDR)
+	name = strings.Replace(cidr, ".", "-", -1)
```

Comment:
Also here - need to agree the name calculation. This is a hack.

Reply:
Don't we already have a function that converts a CIDR to a k8s name (for storing IPPools?)

Reply:
Yep, looks like it basically does the same thing as this. I've updated the code to use that now.